alternator/conditions.cc

```diff
@@ -442,10 +451,11 @@ static bool verify_expected_one(const rjson::value& condition, const rjson::valu
         case comparison_operator_type::CONTAINS:
             verify_operand_count(attribute_value_list, exact_size(1), *comparison_operator);
             return check_CONTAINS(got, (*attribute_value_list)[0]);
-        default:
-            // FIXME: implement all the missing types, so there will be no default here.
-            throw api_error("ValidationException", format("ComparisonOperator {} is not yet supported", *comparison_operator));
+        case comparison_operator_type::NOT_CONTAINS:
+            verify_operand_count(attribute_value_list, exact_size(1), *comparison_operator);
+            return check_NOT_CONTAINS(got, (*attribute_value_list)[0]);
         }
+        __builtin_unreachable();
```

Comment:
It's unclear what this builtin_unreachable does, or what we expect it to do. If we want it to abort(), we could call abort() ourselves. If we want to throw an exception, we could do that explicitly.
Unfortunately I think the safest and cleanest approach is to leave the "default" (sorry about the misleading comment), just throw an exception (e.g., std::logic_error, not an api_error) - this should never happen, but if it does (e.g., in case of memory corruption or unexpected change in the code) at least we behave sensibly and lost just the one request and don't crash the entire cluster.

Reply:
It shuts up the compiler that fails to see that there is no way for the execution path to reach here (sans corrupt enum value) as all valid values are covered by the if. By adding a `default` case you throw away the free reminder from the compiler when a new enum value is added in the future. If you are worried about corrupt values than this `__builtin_unreachable()` can be replaced by `std::abort()`.

Reply:
You're right about the "free reminder" when default is lacking. Yes, I think we should replace __builtin_unreachable() by something more explicit - either std::abort(), or better yet something like "throw std::logic_error" or something (no need to crash the machine and possibly the entire cluster if we have this bug).

Reply:
A potential problem with throwing is that another compiler can start warning that it's dead code, and with our `-Werror` policy (with which I agree), it will result in Scylla not being compilable. We already use `__builtin_unreachable` both in Scylla and Seastar in some places, so perhaps it's alright for it to stay here too. After all, **if** we assume `-Werror`, adding an enum value unhandled in this switch will not compile, so the code is in fact guaranteed to be unreachable.

Reply:
It's not **really** guaranteed to be unreachable, if the enum's underlying variable gets corrupted, it may contain the number 42, which isn't any of the known values for the enum, in which case, the function needs to do *something*. I don't know what __builtin_unreachable does in this case, it probably calls abort() but this isn't defined. I thought it's better to throw in this case - it's an internal error, yes, but it doesn't mean we need to abort the entire server just because we had a problem fulfilling one command.

Reply:
ok, makes sense

Reply:
`__builtin_unreachable` doesn't do anything IIRC, it just tells the compiler: "this is unreachable, pinky promise, don't complain to me". See https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005funreachable.

Reply:
It causes the compiler to shut up and not generate a return statement.

Reply:
Replaced with `throw std::logic_error`.